### DISCUSSION TITLE
Using custom MENDER_DEVICE_TYPE

### DISCUSSION THREAD

**Discussion by lhrvth** [OE4T/tegra-demo-distro#95] [Q&A] [answered]
Hi,
I'm on the master branch and trying to build a tegrademo-mender image using a custom Mender device type for my Jetson AGX Xavier board. The device type is set in tegrademo-mender.conf (this is the only change I made to the demo):
```
DISTRO_VERSION = "${DISTRO_VERSION_BASE}+snapshot"

### Additions for mender ###
MENDER_DEVICE_TYPE = "foobar-agx"

INHERIT += "mender-full distro_layer_buildinfo"
```
The build is started as follows:
```
$ . ./setup-env -m jetson-agx-xavier-devkit -d tegrademo-mender
$ bitbake core-image-minimal
```
Unfortunately, the image creation fails at some point with the following error message:
```
ERROR: linux-tegra-4.9.201+gitAUTOINC+c68230c5b4-r0 do_bundle_initramfs: Could not find any /mnt/build/tegrademo-mender-
master/tmp/deploy/images/jetson-agx-xavier-devkit/tegra-minimal-initramfs-jetson-agx-xavier-devkit.cpio{.gz|.lz4|.lzo|.lzma|.xz)
for bundling; INITRAMFS_IMAGE_NAME might be wrong.
```
I haven't dug into the code deep enough yet but it seems that the value of the MENDER_DEVICE_TYPE variable isn't used consistently during the build/config process.
The MENDER_DEVICE variables are set correctly (except the last one, it makes no sense to me):
```
$ bitbake -e core-image-minimal > env.log
$ grep ^MENDER_DEVICE env.log 
MENDER_DEVICE_TYPE="foobar-agx"
MENDER_DEVICE_TYPES_COMPATIBLE="foobar-agx jetson-xavier"
MENDER_DEVICE_TYPES_COMPATIBLE_DEFAULT="foobar-agx"
MENDER_DEVICE_TYPE_DEFAULT="jetson-agx-xavier-devkit"
MENDER_DEVICE_TYPE_DEFAULT_beaglebone-yocto_mender-grub="jetson-agx-xavier-devkit-grub"
```
but the INITRAMFS variables are probably not:
```
$ grep ^INITRAMFS_ env.log 
INITRAMFS_FSTYPES="cpio.gz"
INITRAMFS_IMAGE="tegra-minimal-initramfs"
INITRAMFS_IMAGE_BUNDLE="1"
INITRAMFS_LINK_NAME="initramfs-jetson-agx-xavier-devkit"
INITRAMFS_MAXSIZE="131072"
INITRAMFS_NAME="initramfs--1.0-r0-jetson-agx-xavier-devkit-20210502115520"
```
and that leads to the following output:
```
$ ls -1 tmp/deploy/images/jetson-agx-xavier-devkit/
bootlogo-jetson-agx-xavier-devkit-1.0-r0.blob
bootlogo-jetson-agx-xavier-devkit.blob
cboot-jetson-agx-xavier-devkit-32.5.0-r0.bin
cboot-jetson-agx-xavier-devkit.bin
mender-state-scripts
tegra-minimal-initramfs-foobar-agx-20210502115734.cpio.gz
tegra-minimal-initramfs-foobar-agx-20210502115734.manifest
tegra-minimal-initramfs-foobar-agx-20210502115734.testdata.json
tegra-minimal-initramfs-foobar-agx.cpio.gz
tegra-minimal-initramfs-foobar-agx.manifest
tegra-minimal-initramfs-foobar-agx.testdata.json
tos-jetson-agx-xavier-devkit-32.5.1-r0.img
tos-jetson-agx-xavier-devkit.img
```
Not sure if this is a 'bug' or I am missing something. Any advice? Thanks!

**Comment by dwalkes** [comment]
> The MENDER_DEVICE variables are set correctly (except the last one, it makes no sense to me):

Me either, but I doubt it's related.

> but the INITRAMFS variables are probably not:

It looks like those are coming from poky at http://git.yoctoproject.org/cgit/cgit.cgi/poky/tree/meta/classes/kernel.bbclass?h=dunfell#n28 while the `IMAGE_NAME` is set based on `MENDER_DEVICE_TYPE` at https://github.com/mendersoftware/meta-mender/blob/master/meta-mender-core/classes/mender-setup.bbclass#L236.  I don't understand how the setting for `IMAGE_NAME` would be related to the generation of the initramfs but it seems like this is what must be happening.

> Not sure if this is a 'bug' or I am missing something.

I'm guessing you might be the first one trying to use a `MENDER_DEVICE_TYPE` which doesn't mach `MACHINE` with poky + tegra + meta-mender-community.  So it's probably a bug but not obvious to me where on first glance.

**Comment by madisongh** [comment, marked as answer]
I had trouble with this as well. The problem is that mender-core changes the definition of `IMAGE_NAME` and `IMAGE_LINK_NAME` globally, replacing `${MACHINE}` with `${MENDER_DEVICE_TYPE}`. However, there are some places where built images get referenced by other recipes - `kernel.bbclass`, which Dan mentioned, being a common one, since initramfs images are often bundled into the kernel. I worked around the issue by setting those two variables back to their `${MACHINE}`-based values explicitly in my initramfs image recipes:
```
# Mender changes these globally, but for the initramfs image we
# need them to be the normal MACHINE-based names.
IMAGE_NAME = "${IMAGE_BASENAME}-${MACHINE}-${IMAGE_VERSION_SUFFIX}"
IMAGE_LINK_NAME = "${IMAGE_BASENAME}-${MACHINE}"
```
Adding a `tegra-minimal-initramfs.bbappend` to do this in your setup should work.

I'm not sure why Mender makes that change in the first place, but I suspect it's because they want built Mender artifacts to include the device type name instead of the MACHINE name, to match the device type value that gets embedded in the artifact header.

**Reply by lhrvth** [reply]
Well, then I'm not the first one trying to use a custom device type :) That's definitely a great workaround, thanks for the tip, Matt!
Thank you both for your help and quick responses!
